sklearn/utils/__init__.py

```diff
@@ -343,21 +358,27 @@ def _safe_indexing(X, indices, *, axis=0):
     if axis == 0 and indices_dtype == "str":
         raise ValueError("String indexing is not supported with 'axis=0'")
 
-    if axis == 1 and X.ndim != 2:
+    if axis == 1 and hasattr(X, "ndim") and X.ndim != 2:
         raise ValueError(
             "'X' should be a 2D NumPy array, 2D sparse matrix or pandas "
             "dataframe when indexing the columns (i.e. 'axis=1'). "
             "Got {} instead with {} dimension(s).".format(type(X), X.ndim)
         )
 
-    if axis == 1 and indices_dtype == "str" and not hasattr(X, "loc"):
+    if (
+        axis == 1
+        and indices_dtype == "str"
+        and not (_is_pandas_df(X) or _use_interchange_protocol(X))
+    ):
         raise ValueError(
             "Specifying the columns using strings is only supported for "
             "pandas DataFrames"
         )
 
     if hasattr(X, "iloc"):
```

Comment:
Maybe we could be more explicit here using `_is_pandas_df`?

Reply:
Changing this would break some test that is unrelated to this PR such as:

https://github.com/scikit-learn/scikit-learn/blob/714c50092b1bfafad2b568a07d635d5db1301d58/sklearn/model_selection/tests/test_split.py#L1435-L1438

Reply:
OK I see.